lib/matplotlib/backends/backend_gtk4.py

```diff
@@ -0,0 +1,683 @@
+        FigureCanvasBase.pick(self, mouseevent)
+
+    def destroy(self):
+        self.close_event()
+
+    def set_cursor(self, cursor):
+        # docstring inherited
+        self.set_cursor_from_name(_mpl_to_gtk_cursor(cursor))
+
+    def scroll_event(self, controller, dx, dy):
+        FigureCanvasBase.scroll_event(self, 0, 0, dy)
+        return True
+
+    def button_press_event(self, controller, n_press, x, y):
+        # flipy so y=0 is bottom of canvas
+        y = self.get_allocation().height - y
+        FigureCanvasBase.button_press_event(self, x, y,
+                                            controller.get_current_button())
+        self.grab_focus()
+
+    def button_release_event(self, controller, n_press, x, y):
+        # flipy so y=0 is bottom of canvas
+        y = self.get_allocation().height - y
+        FigureCanvasBase.button_release_event(self, x, y,
+                                              controller.get_current_button())
+
+    def key_press_event(self, controller, keyval, keycode, state):
+        key = self._get_key(keyval, keycode, state)
+        FigureCanvasBase.key_press_event(self, key)
+        return True
+
+    def key_release_event(self, controller, keyval, keycode, state):
+        key = self._get_key(keyval, keycode, state)
+        FigureCanvasBase.key_release_event(self, key)
+        return True
+
+    def motion_notify_event(self, controller, x, y):
+        # flipy so y=0 is bottom of canvas
+        y = self.get_allocation().height - y
+        FigureCanvasBase.motion_notify_event(self, x, y)
+
+    def leave_notify_event(self, controller):
+        FigureCanvasBase.leave_notify_event(self)
+
+    def enter_notify_event(self, controller, x, y):
+        # flipy so y=0 is bottom of canvas
+        y = self.get_allocation().height - y
+        FigureCanvasBase.enter_notify_event(self, xy=(x, y))
+
+    def resize_event(self, area, width, height):
+        dpi = self.figure.dpi
+        self.figure.set_size_inches(width / dpi, height / dpi, forward=False)
+        FigureCanvasBase.resize_event(self)
+        self.draw_idle()
+
+    def _get_key(self, keyval, keycode, state):
+        unikey = chr(Gdk.keyval_to_unicode(keyval))
+        key = cbook._unikey_or_keysym_to_mplkey(
+            unikey,
+            Gdk.keyval_name(keyval))
+        modifiers = [
+            (Gdk.ModifierType.CONTROL_MASK, 'ctrl'),
+            (Gdk.ModifierType.ALT_MASK, 'alt'),
+            (Gdk.ModifierType.SHIFT_MASK, 'shift'),
+            (Gdk.ModifierType.SUPER_MASK, 'super'),
+        ]
+        for key_mask, prefix in modifiers:
+            if state & key_mask:
+                if not (prefix == 'shift' and unikey.isprintable()):
+                    key = f'{prefix}+{key}'
+        return key
+
+    def _draw_rubberband(self, rect):
+        self._rubberband_rect = rect
+        # TODO: Only update the rubberband area.
+        self.queue_draw()
+
+    def draw_func(self, drawing_area, ctx, width, height):
```

Comment:
Keep this private?  (this makes is slightly clearer that 3rd party renderers (i.e., mplcairo.gtk) can keep on overriding on_draw_event as in gtk3, not something else)